### PR TITLE
OTA-1393: status: recognize multi-arch in node phase (2)

### PR DESCRIPTION
**Pull request by hongkailiu**
It turns out that MCO [annotates](https://github.com/openshift/oc/pull/1928/files#diff-a8a4caed2617385d6b33f8895922756270b8eb9fe0a679d5afac9de8471e9357R163) a node and updates it and only then moves on the next node.

This means for example, the worker pool with 3 nodes has always `COMPLETION=67% (2/3)` during migration.

I thought about use `mcp.status.updatedMachineCount/mcp.status.machineCount` for the migration case. 
With [the mcp in our example](https://github.com/openshift/oc/blob/7aebefa5ea9cfac517479b66531f187536db9593/pkg/cli/admin/upgrade/status/examples/to-multi-arch-mcp.yaml#L243-L247), `COMPLETION=33% (1/3)` instead of `COMPLETION=67% (2/3)`.
It would work ONLY for the case without the customized pools. To fix it properly, we might have to resort to the changes on MCO.

`N/A` does not sound ideal but an admin might not be interested in the progress of existing worker nodes restarts as it is irrelevant to the goal of migrating a cluster to multi-arch: create a new pool of nodes with a secondary arch.
As soon as the control planes upgrade completes, the cluster is ready for that.

---
With the idea of https://github.com/openshift/oc/pull/1933#discussion_r1871941964, we fixed the node phase, assessment, and the completion.

---
We found a limitation that the status command claims "not updating" while the data plane update is still running (after the control plane update is complete). See [the discussion](https://github.com/openshift/oc/pull/1933#discussion_r1881155093).

/cc @petr-muller @DavidHurta

### CODE REVIEW

File: pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output
Comment (DavidHurta) [inline pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output:22]: This whole issue nicely demonstrates that the current evaluation of whether a node `isUpdated` has flaws.

Let's try to fix the underlying logic.

> It turns out that MCO [annotates](https://github.com/openshift/oc/pull/1928/files#diff-a8a4caed2617385d6b33f8895922756270b8eb9fe0a679d5afac9de8471e9357R163) a node and updates it and only then moves on the next node.

That is true. However, a `MachineConfigPool` resource itself holds the `MachineConfig` it wants to apply for its nodes. We could try to use that!

> .spec.configuration
> The targeted MachineConfig object for the machine config pool.

I have not delved deeper into the API, but it looks promising! 

We can use the field to enhance our [current](https://github.com/openshift/oc/pull/1928/files#diff-a8a4caed2617385d6b33f8895922756270b8eb9fe0a679d5afac9de8471e9357R163) logic.
Something like:
```
isUpdated := foundCurrent && isLatestUpdateHistoryVersionEqualTo(cv.Status.History, currentVersion) &&
	// The following condition is to handle the multi-arch migration because the version number stays the same there
	(!ok || (node.Annotations[mco.CurrentMachineConfigAnnotationKey] == desiredConfig && desiredConfig == pool.Spec.Configuration.Name))
```
Feel free to check for the updated examples using the field at https://github.com/DavidHurta/oc/commit/88f8853ab562f5285e4979e93ea87eb73466273b. The completion column now looks correct!

However, the other things are not so well. This is because the `isUpdating` was not updated up to speed for the checking of `MachineConfigs`. We can now update it so it knows about versions, `MachineConfigs`, and `MachineConfigPools`! Everything looks great, IMO! Feel free to check the examples and the updated logic at https://github.com/DavidHurta/oc/commit/ab5ede5a760e810c84110ff6324769a510dabe63.

The `isUpdating` was updated to:
```
isUpdating := !isUpdated && foundCurrent && foundDesired && isLatestUpdateHistoryVersionEqualTo(cv.Status.History, desiredVersion) &&
	(!ok || (node.Annotations[mco.CurrentMachineConfigAnnotationKey] != desiredConfig && desiredConfig == pool.Spec.Configuration.Name))
```

I have manually checked the nodes section in my commit against the resources, and it seems to be correct. Could you have a look at the API mentioned and these comment ideas to see if they help you solve the issue?

**Review by DavidHurta — CHANGES_REQUESTED** [review]

File: pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output
Comment (DavidHurta) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output:22]: I am afraid I do not understand the `It would work ONLY for the case without the customized pools.` sentence. It is possible my comment is irrelevant to that use case. Could you elaborate on the sentence?

Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: > `.spec.configuration`

Great catch.
I twisted condition a bit for the readablity.

> Could you elaborate on the sentence?

With https://github.com/openshift/oc/pull/1825/files we removed the nodes in custom pools from the general workers. 
However, `MCP/worker` still displays the workers in custom pool. Thus, I assume that the numbers in its status counts those nodes as well.
In with help of `.spec.configuration`, we do not hit this issue as we count everything on our own (as you said, irrelevant to us now).

Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: https://asciinema.org/a/Q3QGWshS2wVX8ksuhEupmTsv1 looks great to me!

For some reason, one machine is in trouble becoming a worker node. But we can ignore that.

```
oc get machine -n openshift-machine-api
NAME                                                PHASE          TYPE         REGION      ZONE         AGE
ci-ln-hicmvjt-76ef8-77sz6-master-0                  Running        m6a.xlarge   us-west-2   us-west-2b   60m
ci-ln-hicmvjt-76ef8-77sz6-master-1                  Running        m6a.xlarge   us-west-2   us-west-2d   60m
ci-ln-hicmvjt-76ef8-77sz6-master-2                  Running        m6a.xlarge   us-west-2   us-west-2b   60m
ci-ln-hicmvjt-76ef8-77sz6-worker-us-west-2b-k5h44   Running        m6a.xlarge   us-west-2   us-west-2b   56m
ci-ln-hicmvjt-76ef8-77sz6-worker-us-west-2b-xxzj9   Running        m6a.xlarge   us-west-2   us-west-2b   56m
ci-ln-hicmvjt-76ef8-77sz6-worker-us-west-2d-df6mw   Provisioning                                         10s
```

Comment (DavidHurta) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: However, it is worth noting that we are changing the logic of "_is cluster updating?_" at https://github.com/openshift/oc/blob/d5b4a430f7e3307d88cd99012daa60a2f491bd50/pkg/cli/admin/upgrade/status/status.go#L305-L308

Note that it was already changed in https://github.com/openshift/oc/pull/1928.

I do not know how often MCO rolls out new `MachineConfig` across nodes. Is it only when a new release payload is being applied? If not, the current command is described as:

```
$ OC_ENABLE_CMD_UPGRADE_STATUS=true ./oc adm upgrade status --help
Display the status of current cluster version updates.
```

would show that the current cluster version is being upgraded even though the MCO is just rolling out a new `MachineConfig` due to some factors (I am not that familiar with `MachineConfigs`) and not due to the cluster version being upgraded. The nodes would be technically being updated but it would not be exactly a cluster version update. I am not saying whether this is good or bad, I am not sure. **It is worth noting.**

However, I do not know how to capture this specific use case otherwise. Multi migration can't be recognized using versions, and a control plane can be already updated with the `ClusterVersion` reporting that an update is `Completed`. And I do not know how to check whether nodes are being migrated to multi-arch vs. MCO is rolling out a slightly modified `MachineConfig`. This would need additional exploration. So it looks like we will either report a cluster is upgraded even when a release is not being applied, or we will limit the output when migrating, or explore some more options?

Comment (DavidHurta) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: We can also update the description and thus the definition of the `oc adm upgrade status` subcommand.
```diff
$ OC_ENABLE_CMD_UPGRADE_STATUS=true ./oc adm upgrade status --help
- Display the status of current cluster version updates.
+ Display the status of current updates.
```

Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: > we are changing the logic of "is cluster updating?"

> I do not know how often MCO rolls out new MachineConfig across nodes.

My claim is that "it is fine to change it". This is my thinking.
With this PR and PR#1928, the conditions for a node to be updated are harder to meet than before.
It means a previously updated cluster could become outdated now.

As we noticed, cluster updates generate new MCs. There are occassions as well, for example, [changing something on the nodes](https://docs.openshift.com/container-platform/4.17/machine_configuration/machine-configs-configure.html). In those cases, I would assume MCP also reports the pool is being updated.

The changes with the 2 PRs make us closer to MCO.
In theory, as long as we are aligned our concepts with MCO's components (such as MCP in this case), we should be fine.

> We can also update the description and thus the definition of the oc adm upgrade status subcommand.

Maybe we can still keep the "cluster" and remove only the "version"? (Not relevant to this PR: "update" as singleton?)

```diff
$ OC_ENABLE_CMD_UPGRADE_STATUS=true ./oc adm upgrade status --help
- Display the status of current cluster version updates.
+ Display the status of current cluster update.
```

Comment (DavidHurta) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: > So the non-version related updates are there already even without the multi-arch migration.

Yes, that is correct. The non-version related updates, however, were not previously shown because this was gated by the version check being the only check. However, that does not work for multi-arch migration, which is a valid cluster version update. 

`Display the status of the current cluster update.` is probably more grammatically correct.

I am happy to approve the current logic (still need to check the syntax). However, we should after the PR is merged, check the command against non-version related updates to check things like "Does the output makes sense?" "Do we display correctly the target version?". I would guess that there is a code working with the assumption that it's working only in cluster version updates.

/cc @petr-muller 
as we are discussing the usage of the command

Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: > However, we should after the PR is merged, check the command against non-version related updates to check things like "Does the output makes sense?" "Do we display correctly the target version?"

Let me do it before the merge.
I can even do the comparison with 1 cluster with 2 oc (with and without this PR and PR#1928 )

Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: > The non-version related updates, however, were not previously shown (at least I think) because this was gated by the version check being the only check.

You are correct because I saw the diff. in my testing, by applying the following to the cluster.

```console
cat << EOF > 99-worker-realtime.yaml
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: "worker"
  name: 99-worker-realtime
spec:
  kernelType: realtime
EOF
```

The oc-cli with `Client Version: 4.17.5` shows "The cluster is not updating." while the oc-cli built with my PR shows all the fancy output.

> And I do not know how to check whether nodes are being migrated to multi-arch vs. MCO is rolling out a slightly modified MachineConfig.

So I stepped back to the above and we actually have such a function. See [f238d35](https://github.com/openshift/oc/pull/1933/commits/f238d35925f6be81d5b133c6d73a2f41a31b7274).

Comment (petr-muller) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: /hold

well spotted @DavidHurta 
I'll check the convo later today - I know the PR is now probably good to go ("So I stepped back to the above and we actually have such a function.") but let me take a look :+1:

Comment (petr-muller) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: The MCO-is-just-rolling-out-changes vs MCO-is-actually-rolling-out-changes-related-to-version-update dichotomy is a bit of a pain for a long time. Similarly like for CVO, an actual version update of a node is simply a slightly special case of MCO reconciliation. This is true for the system, but it is _very_ counterintuitive for (at least some) customers; they simply expect the "upgrade" to be something else (=bigger) than simply rolling out changes, but it is not. One annoying part is that MCO is rolling out changes anytime there is a new node. That's very often in autoscaling clusters.

The web console actually contains a similar broader behavior: it will display an "updating" progress bar when there is simply a mco change rollout in progress. This is known to confuse customers, sometimes they even think the cluster started "updating" on its own because they see this progress bar.

**Therefore I would really like if we would only show any update-related output, only if if it is related to an actual version update, and not otherwise.**

Comment (petr-muller) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: /hold cancel

Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: > This is known to confuse customers, sometimes they even think the cluster started "updating" on its own because they see this progress bar.

There was a time when It confused me.

> ... only if if it is related to an actual version update, and not otherwise. The single exception is the multi arch migration.

I am thinking if we can avoid regression. Let me if I can get unit tests to achieve this.

File: pkg/cli/admin/upgrade/status/status.go
Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/status.go]: Did a little rewording to include the migration case.

File: pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output
Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/to-multi-arch.detailed-output]: Done with a separate commit.

**Review by petr-muller — APPROVED** [review]
LGTM, thanks

/hold

@DavidHurta holding for you to check if your concerns are addressed

**Comment by DavidHurta** [comment]
Thanks, I will check the PR before the end of the day.

File: pkg/cli/admin/upgrade/status/examples/no-version-upgrading-cv.yaml
Comment (DavidHurta) [reply pkg/cli/admin/upgrade/status/examples/no-version-upgrading-cv.yaml:114]: I suppose a user can also run `oc adm upgrade --clear` after migrating to multi-arch. This would mean that we would lose the `ClusterVersion.spec.desiredUpdate` field, which contains the `architecture` field :sweat_smile: However, the previous case is concerning me more.

Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/no-version-upgrading-cv.yaml]: `Progressing=True` is added into the `updated` conditions. Your case (mcp reconciliation on a migrated cluster) will be handled correctly: annotation checks for migration happens only when migration is in progress.

The manifests are taken from a new cluster-bot cluster. 

>  we would lose the `ClusterVersion.spec.desiredUpdate` field

Removed the condition about `ClusterVersion.spec.desiredUpdate`.

File: pkg/cli/admin/upgrade/status/workerpool.go
Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/workerpool.go:171]: My thinking is `(!annotationMatched && desiredAnnotationUpdated)` is implicitly `true` (provided `!isUpdated && foundCurrent && foundDesired && isLatestUpdateHistoryVersionEqualTo(cv.Status.History, desiredVersion)` is true) for an updating node during a normal version update. (It is necessary only for multi-arch migration tho.)
Otherwise, one of the following conditions is true:
- `node.Annotations[mco.CurrentMachineConfigAnnotationKey]==node.Annotations[mco.DesiredMachineConfigAnnotationKey]` which would contradicts to `!isUpdated`.
- `desiredConfig != pool.Spec.Configuration.Name`. I bet MCO always sets those two up during update even if it is an optional field.

File: pkg/cli/admin/upgrade/status/examples/no-version-upgrading-cv.yaml
Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/examples/no-version-upgrading-cv.yaml]:
```console
    --clear=false:
	If an upgrade has been requested but not yet downloaded, cancel the update. This has no
	effect once the update has started.
```

Would we still lose `ClusterVersion.spec.desiredUpdate` after "the update has started"?
If the field is still there, we do not need to "Removed the condition about ClusterVersion.spec.desiredUpdate.".
Otherwise, "This has no effect" is not accurate.

Let me test it.

**Comment by petr-muller** [comment]
@hongkailiu unless you really want to, I don't think it's necessary to pass `okd-scos-e2e-aws-ovn` job unless it indicates a problem that seems related to your PR. It is not a required job.

**Comment by hongkailiu** [comment]
> @hongkailiu unless you really want to, I don't think it's necessary to pass `okd-scos-e2e-aws-ovn` job unless it indicates a problem that seems related to your PR. It is not a required job.

I was trying blindly as the PR still awaits reviews.
The job was green once earlier so I thought it might be just flaky.
The error from the failing job indicates troubles connecting the API server of the ephemeral cluster which I feel unrelated to my PR.
I will stop trying now.

**Comment by petr-muller** [comment]
I am only skimming the revisions now - I assume @DavidHurta is the final approver now :+1:

File: pkg/cli/admin/upgrade/status/status.go
Comment (hongkailiu) [reply pkg/cli/admin/upgrade/status/status.go:290]: > Acknowledge limitations and be ok with them.

I prefer to this one too.
I admit that data plane update is ignored for the technical difficulty.
Arguable, it might be an even better behaviour. See my comment.

https://github.com/openshift/oc/pull/1933/files#diff-a7caf221970b9d8c0d7e98b344e66c2f5d88490c787d696f8162b2d62630bc05R285-R290

"Ideally MCO do not have to reboot the existing machines for the machine-config changes caused by the multi-arch migration" is discussed https://redhat-internal.slack.com/archives/C02CZNQHGN8/p1733410304411049

**Review by DavidHurta — APPROVED** [review]
Looks great, thanks for the work and the discussions! :raised_hands: 

Note: Please update the PR's description.

**Pull request merged**
